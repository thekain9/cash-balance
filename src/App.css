@import url('https://fonts.googleapis.com/css2?family=Montserrat:wght@400;600&family=Roboto:wght@100;400&display=swap');

.App {
  display: flex;
  align-items: center;
  justify-content: center;
}


body {
  margin: 0; /* Remove default margin */
  display: flex;
  justify-content: center; /* Center horizontally */
  align-items: center; /* Center vertically */
  min-height: 100vh;
  background-color: whitesmoke;
  
}

h1 {
  font-family: 'Montserrat', sans-serif;
  font-weight: 600;
  font-size: 5rem;
}


.buttons {
  display: flex;
  padding: 3rem;
  justify-content: center;
}

.buttons button {
  font-family: 'Montserrat', sans-serif;
  font-weight: 600;
  font-size: 1rem;
  padding: 1.5rem;
  margin: 2rem;
  width: 10rem;
  border-radius: 2rem;
  background-color: #61dafb;
}

.form {
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: 1.5rem;
  
}

.form label {
  display: flex;
  flex-direction: column;
  font-size: 1rem;
  padding: 1.5rem;
  justify-content: center;
  font-weight: 600;
  font-family: 'Montserrat', sans-serif;
  align-items: center;
}

input {
  font-family: 'Montserrat', sans-serif;
  font-weight: 600;
  border-radius: 2rem;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
