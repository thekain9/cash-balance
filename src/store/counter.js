// Import the necessary function from the Redux Toolkit library
import { createSlice } from "@reduxjs/toolkit";

// Create a slice for the counter with initial state
export const counterSlice = createSlice({
    name: "counter", // Name of the slice

    initialState: {
        value: 0, // Initial state value for the counter
    },

    // Reducer functions for various actions
    reducers: {
        // Reducer function for withdrawing from the balance
        withdraw: (state, action) => {
            state.value -= action.payload; // Decrease the balance by the action payload
        },

        // Reducer function for depositing to the balance
        deposit: (state, action) => {
            state.value += action.payload; // Increase the balance by the action payload
        },

        // Reducer function for adding interest to the balance
        addInterest: (state) => {
            state.value = parseFloat((state.value * 1.05).toFixed(2)); // Increase by 5% and round to 2 decimals
        },
        
        charges: (state) => {
            state.value = parseFloat((state.value * 0.85).toFixed(2)); // Decrease by 15% and round to 2 decimals
        },
    },
});

// Export individual action creators
export const { withdraw, deposit, addInterest, charges } =
    counterSlice.actions;

// Export the reducer function for the counter slice
export default counterSlice.reducer;
