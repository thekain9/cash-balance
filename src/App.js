import { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { withdraw, deposit, addInterest, charges } from './store/counter';
import './App.css';

function App() {
  const [userInput, setUserInput] = useState('');
  const dispatch = useDispatch();

  const handleOperation = (operation) => {
    const amount = Number(userInput);

    // Check if the user input is a valid number
    if (!isNaN(amount)) {
      switch (operation) {
        case 'withdraw':
          dispatch(withdraw(amount));
          break;
        case 'deposit':
          dispatch(deposit(amount));
          break;
        case 'addInterest':
          dispatch(addInterest(amount));
          break;
        case 'charges':
          dispatch(charges(amount));
          break;
        default:
          break;
      }

      setUserInput(''); // Reset input after performing operation
    }
  };

  const count = useSelector((state) => state.counter.value);

  // Fetch current time from an API
  const [fetchedTime, setFetchedTime] = useState(null);

  // Fetch current time from an API and set it in the state
  const fetchTime = async () => {
    try {
      const response = await fetch('https://worldtimeapi.org/api/ip');
      const data = await response.json();
      setFetchedTime(data.datetime);
    } catch (error) {
      console.error('Error fetching time:', error);
    }
  };

  return (
    <div className="page">
      <div className="App">
        <h1>Â£{count}</h1>
      </div>
      <div className="buttons">
        <button onClick={() => handleOperation('withdraw')}>Withdraw</button>
        <button onClick={() => handleOperation('deposit')}>Deposit</button>
        <button onClick={() => handleOperation('addInterest')}>Add Interest</button>
        <button onClick={() => handleOperation('charges')}>Charges</button>
        <button onClick={fetchTime}>Date and time</button>
      </div>
  
      <form className="form">
        <label>
          Value:
          <input
            type="text"
            name="value"
            onChange={(e) => setUserInput(e.target.value)}
            value={userInput}
          />
        </label>
      </form>
  
      {/* Display fetched time */}
      <h1>Date and time: {fetchedTime}</h1>
      
    </div>
  );
  }

export default App;
