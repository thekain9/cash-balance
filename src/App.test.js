import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer'; // Import renderer for snapshot testing
import { Provider } from 'react-redux';
import store from './store/store'; 
import App from './App';


describe('App component', () => {
  // Snapshot test for initial rendering
  test('renders initial value (snapshot)', () => {
    // Render the App component wrapped in the Redux Provider with the store
    const tree = renderer.create(
      <Provider store={store}>
        <App />
      </Provider>
    );

    // Convert the rendered component tree to JSON for snapshot comparison
    const treeJSON = tree.toJSON();

    // Expect the rendered tree to match the stored snapshot
    expect(treeJSON).toMatchSnapshot();
  });



  // Test: Verify that fetch data works correctly
  test('fetch data works correctly', async () => {
    // Mock the fetch function to return a predefined response
    const mockResponse = { datetime: '2023-08-30T12:34:56' };
    jest.spyOn(global, 'fetch').mockResolvedValue({
      json: jest.fn().mockResolvedValue(mockResponse),
    });

    render(
      <Provider store={store}>
        <App />
      </Provider>
    );

    const fetchButton = screen.getByText('Date and time');
    fireEvent.click(fetchButton);

    // Wait for the response and check if the fetched time is displayed
    const fetchedTimeElement = await screen.findByText('Date and time: 2023-08-30T12:34:56');
    expect(fetchedTimeElement).toBeInTheDocument();

    // Clean up the mock after the test
    global.fetch.mockRestore();
  });

  //-------------------FUNCITONAL TESTS--------------------


  // Test: Verify that the value updates correctly after a deposit
  test('updates value after deposit', () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>
    );

    const depositButton = screen.getByText('Deposit');
    const inputElement = screen.getByLabelText('Value:');
    
    fireEvent.change(inputElement, { target: { value: '100' } });
    fireEvent.click(depositButton);

    const updatedValueElement = screen.getByText('£100');
    expect(updatedValueElement).toBeInTheDocument();
  });

  // Test: Verify that the value updates correctly after a withdraw
  test('updates value after withdraw', () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>
    );

    const withdrawButton = screen.getByText('Withdraw');
    const inputElement = screen.getByLabelText('Value:');
    
    fireEvent.change(inputElement, { target: { value: '50' } });
    fireEvent.click(withdrawButton);

    const updatedValueElement = screen.getByText('£50'); // Assuming withdraws can result in negative balance

    expect(updatedValueElement).toBeInTheDocument();
  });

  // Test: Verify that the value updates correctly after adding interest
  test('updates value after adding interest', () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>
    );

    const addInterestButton = screen.getByText('Add Interest');

    fireEvent.click(addInterestButton);

    const updatedValueElement = screen.getByText('£52.5'); // Assuming initial value was £50, and 5% interest

    expect(updatedValueElement).toBeInTheDocument();
  });

  // Test: Verify that the value updates correctly after applying charges
  test('updates value after applying charges', () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>
    );

    const chargesButton = screen.getByText('Charges');

    fireEvent.click(chargesButton);

    const updatedValueElement = screen.getByText('£44.63'); // Assuming initial value was £50, and 15% charges

    expect(updatedValueElement).toBeInTheDocument();
  });
});
