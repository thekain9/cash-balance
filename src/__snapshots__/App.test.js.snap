// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders initial value (snapshot) 1`] = `
<div
  className="page"
>
  <div
    className="App"
  >
    <h1>
      Â£
      0
    </h1>
  </div>
  <div
    className="buttons"
  >
    <button
      onClick={[Function]}
    >
      Withdraw
    </button>
    <button
      onClick={[Function]}
    >
      Deposit
    </button>
    <button
      onClick={[Function]}
    >
      Add Interest
    </button>
    <button
      onClick={[Function]}
    >
      Charges
    </button>
    <button
      onClick={[Function]}
    >
      Date and time
    </button>
  </div>
  <form
    className="form"
  >
    <label>
      Value:
      <input
        name="value"
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
  </form>
  <h1>
    Date and time: 
  </h1>
</div>
`;
